ecommerce/processes/lib/processes/reservation_process.rb

```diff
@@ -28,10 +21,29 @@ def call(event)
     private
 
     def reserve_stock(state)
+      unavailable_products = []
       state.order_lines.each do |product_id, quantity|
         command_bus.(Inventory::Reserve.new(product_id: product_id, quantity: quantity))
         state.product_reserved(product_id)
+      rescue Inventory::InventoryEntry::InventoryNotAvailable
+        unavailable_products << product_id
+      end
+
+      if unavailable_products.empty?
+        event = ReservationProcessSuceeded.new(data: { order_id: state.order_id })
+      else
+        release_stock(state)
+        event = ReservationProcessFailed.new(data: { order_id: state.order_id, unavailable_products: unavailable_products })
       end
+      event_store.publish(event, stream_name: stream_name(state.order_id))
+    end
+
+    def order_side_effects(state)
+      event_store
+      .within { yield }
+      .subscribe(to: ReservationProcessFailed) { reject_order(state) }
+      .subscribe(to: ReservationProcessSuceeded) { accept_order(state) }
+      .call
```

Comment:
I think these side effects could be extracted to fully-fledged event handlers. Then the ReservationProcess could have only a single responsibility. This PR is already big so I didn't want to extract it here. Let me know what you think and I will add an issue for this refactor.